File: frontend/src/lib/api/development.ts
import type { z } from 'zod'
import { API_URL } from '#/lib/api/client'
import { ApiError } from '#/lib/api/errors'
import {
  developmentItemCreateSchema,
  developmentItemSchema,
  developmentItemUpdateSchema,
  developmentPlanResponseSchema,
  type DevelopmentItem,
  type DevelopmentPlanResponse,
} from '#/lib/api/developmentSchemas'

// R17 #199 development-plan client. Mirrors the sibling admin.ts feature client:
// its own owner-scoped fetch wrapper (single silent-refresh on 401) plus Zod
// validation of every response against the shared developmentSchemas. Request
// bodies are parsed through the same schemas so `null` (clear) vs. omitted
// (leave unchanged) is preserved end-to-end (D-112).

export type DevelopmentItemCreate = z.infer<typeof developmentItemCreateSchema>
export type DevelopmentItemUpdate = z.infer<typeof developmentItemUpdateSchema>

async function developmentFetch(path: string, options: RequestInit = {}) {
  const headers = new Headers(options.headers || {})
  if (options.body && typeof options.body === 'string' && !headers.has('Content-Type')) {
    headers.set('Content-Type', 'application/json')
  }
  return fetch(`${API_URL}${path}`, {
    ...options,
    headers,
    credentials: 'include',
    signal: AbortSignal.timeout(30_000),
  })
}

async function developmentRequest(path: string, options: RequestInit = {}): Promise<unknown> {
  let response = await developmentFetch(path, options)

  if (response.status === 401 && path !== '/auth/refresh') {
    const refreshResponse = await developmentFetch('/auth/refresh', {
      method: 'POST',
      body: '{}',
    }).catch(() => null)

    if (refreshResponse?.ok) {
      response = await developmentFetch(path, options)
    } else {
      window.dispatchEvent(new CustomEvent('cw:session-expired'))
    }
  }

  // 204 No Content (DELETE) carries no body to parse.
  if (response.status === 204) {
    if (!response.ok) throw new ApiError('Request failed', response.status)
    return undefined
  }

  const parsed = await response
    .clone()
    .json()
    .catch(async () => response.text().catch(() => ''))

  if (!response.ok) {
    const detail =
      typeof parsed === 'string'
        ? parsed
        : typeof parsed === 'object' && parsed && 'detail' in parsed
          ? String((parsed as Record<string, unknown>).detail)
          : undefined
    throw new ApiError(detail || 'Request failed', response.status, detail)
  }

  return parsed
}

export async function getDevelopmentPlan(): Promise<DevelopmentPlanResponse> {
  const data = await developmentRequest('/development-plan', { method: 'GET' })
  return developmentPlanResponseSchema.parse(data)
}

export async function createDevelopmentItem(
  payload: DevelopmentItemCreate,
): Promise<DevelopmentItem> {
  const data = await developmentRequest('/development-plan', {
    method: 'POST',
    body: JSON.stringify(developmentItemCreateSchema.parse(payload)),
  })
  return developmentItemSchema.parse(data)
}

export async function updateDevelopmentItem(
  itemId: string,
  payload: DevelopmentItemUpdate,
): Promise<DevelopmentItem> {
  const data = await developmentRequest(`/development-plan/${itemId}`, {
    method: 'PATCH',
    body: JSON.stringify(developmentItemUpdateSchema.parse(payload)),
  })
  return developmentItemSchema.parse(data)
}

export async function deleteDevelopmentItem(itemId: string): Promise<void> {
  await developmentRequest(`/development-plan/${itemId}`, { method: 'DELETE' })
}

// R17 #201: act on the proposal completion staged (D-113). Confirming makes it
// available to future tailoring/recommendations; declining hard-deletes it,
// leaving no trace in the Evidence Profile, while the item itself stays intact.

export async function confirmDevelopmentEvidence(itemId: string): Promise<DevelopmentItem> {
  const data = await developmentRequest(`/development-plan/${itemId}/confirm-evidence`, {
    method: 'POST',
  })
  return developmentItemSchema.parse(data)
}

export async function declineDevelopmentEvidence(itemId: string): Promise<DevelopmentItem> {
  const data = await developmentRequest(`/development-plan/${itemId}/decline-evidence`, {
    method: 'POST',
  })
  return developmentItemSchema.parse(data)
}

File: backend/app/routers/development.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.auth.security import get_current_user
from app.database import get_db
from app.models.user import User
from app.schemas.development import (
    DevelopmentItemCreate,
    DevelopmentItemResponse,
    DevelopmentItemUpdate,
    DevelopmentPlanResponse,
)
from app.services.development import (
    DevelopmentItemNotFoundError,
    GapClassificationNotFoundError,
    NoEvidenceProposalError,
    confirm_development_evidence,
    create_development_item,
    decline_development_evidence,
    delete_development_item,
    list_development_items,
    update_development_item,
)

router = APIRouter()


@router.get("", response_model=DevelopmentPlanResponse)
def get_plan(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return DevelopmentPlanResponse(items=list_development_items(db, current_user.id))


@router.post("", response_model=DevelopmentItemResponse, status_code=201)
def create_item(
    body: DevelopmentItemCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    try:
        return create_development_item(db, current_user.id, body)
    except GapClassificationNotFoundError as error:
        raise HTTPException(status_code=404, detail="Gap classification not found") from error


@router.patch("/{item_id}", response_model=DevelopmentItemResponse)
def update_item(
    item_id: str,
    body: DevelopmentItemUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    try:
        return update_development_item(db, item_id, current_user.id, body)
    except DevelopmentItemNotFoundError as error:
        raise HTTPException(status_code=404, detail="Development item not found") from error


@router.delete("/{item_id}", status_code=204)
def delete_item(
    item_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    try:
        delete_development_item(db, item_id, current_user.id)
    except DevelopmentItemNotFoundError as error:
        raise HTTPException(status_code=404, detail="Development item not found") from error


@router.post("/{item_id}/confirm-evidence", response_model=DevelopmentItemResponse)
def confirm_evidence(
    item_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Confirm the proposal completion produced (R17 #201, D-113: user-only)."""
    try:
        return confirm_development_evidence(db, item_id, current_user.id)
    except DevelopmentItemNotFoundError as error:
        raise HTTPException(status_code=404, detail="Development item not found") from error
    except NoEvidenceProposalError as error:
        raise HTTPException(status_code=409, detail="No evidence proposal to confirm") from error


@router.post("/{item_id}/decline-evidence", response_model=DevelopmentItemResponse)
def decline_evidence(
    item_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Decline the proposal: hard-deleted, leaving no trace in the profile; the
    completed item itself is left intact (R17 #201, D-113)."""
    try:
        return decline_development_evidence(db, item_id, current_user.id)
    except DevelopmentItemNotFoundError as error:
        raise HTTPException(status_code=404, detail="Development item not found") from error
    except NoEvidenceProposalError as error:
        raise HTTPException(status_code=409, detail="No evidence proposal to decline") from error
